src/lib/krb5/ccache/ccselect.c

```diff
@@ -142,32 +144,55 @@ krb5_cc_select(krb5_context context, krb5_principal server,
             return ret;
     }
 
+    ret = krb5_copy_principal(context, server, &srvcp);
+    if (ret)
+        return ret;
+
+    /* Try to use the fallback host realm for the server if there's no
+     * authoritative realm. */
+    if (krb5_is_referral_realm(&srvcp->realm)) {
+        ret = krb5_get_fallback_host_realm(context, &srvcp->data[1], &fbrealms);
+        if (ret) {
+            krb5_free_principal(context, srvcp);
+            return ret;
+        }
+        if (fbrealms != NULL && fbrealms[0] != NULL) {
+            krb5_free_data_contents(context, &srvcp->realm);
+            srvcp->realm = string2data(fbrealms[0]);
+            free(fbrealms);
```

Comment:
This leaks memory if fbrealms has more than one value.  fbrealms is also leaked if it contains no values, although that shouldn't happen.